### ISSUE TITLE
Aiosmtpd incompatible with O365 service

### ISSUE THREAD

**Issue by cnicodeme** [aio-libs/aiosmtpd#299]
Hi,

When receiving the command "MAIL FROM", the code located at smtp.py is checking for parameters such as SIZE, SMTPUTF8, etc.

The problem is if it receives an unknown paramter, the code fails at [L1278](https://github.com/aio-libs/aiosmtpd/blob/master/aiosmtpd/smtp.py#L1278):

```
        if len(params) > 0:
            await self.push(
                '555 MAIL FROM parameters not recognized or not implemented')
            return
```

By using this behavior, which is repeated for the RCPT command ([L1319](https://github.com/aio-libs/aiosmtpd/blob/master/aiosmtpd/smtp.py#L1319)), there is absolutely no way to accept an email that contains parameters that AIOSmtpd handles, even valid parameters.

The issue is that service, like O365, seems to send the `MAIL FROM` command with parameters that are unknown to AIOSmtpd, causing the command to be refused and the email never delivered.

I believe that the parameters sent by O365 are valid and should be either accepted, or ignored, **but not refused**.

**Comment by waynew** [comment]
I was thrilled when I saw this issue because I've had issues with sending email from work to my aiostmpd server....

However, I'm not getting the log messages that I thought I would see. Do you have examples of other parameters that o365 tries to send?

**Comment by cnicodeme** [comment]
YES! Following my ticket, I've added tracking to the remaining parameters to know which one was causing issues.

It's `AUTH=<>`

Don't ask me why though :p

But in a broader sense, I don't think it's a good idea to block the command if there are any unknown parameters. I can understand why it would be important to be able to process them all (I mean, they are here for a reason ... well, except O365), but AIOSmtpd does not support all the known parameters so it can easily break with a system sending one of the unsupported parameter.

What I think would be better, is AIOSmtpd calling a hook that handles the parameters once AIOSmptd has finished treating the one it knows. That hook could either send the whole set of parameters, or the remaining one, and then, only if that hook returns a non empty object (the untreated one), then AIOSmtpd could eventually break and return a 550.

If I've understood the RFC correctly, there are a list of official mail-parameters that can be accepted, and they are listed here:
https://www.rfc-editor.org/rfc/rfc6729.html#section-6.2

(again, if I understood correctly, not entirely sure), but again, it would be best to let the developer using AIOSmptd decide what to do in the event an unknown parameter is received.

**Comment by waynew** [comment]
I'm OK with us offering a hook and falling back to failure/rejection.

Interestingly enough I *was* having problems with getting email from outlook on my own server, but that was because I wasn't properly returning `\r\n`, and apparently Outlook is strict about those bytes :upside_down_face:

**Comment by cnicodeme** [comment]
I think it's the best solution. I recently had another notification about extra headers : 

    NOTIFY="SUCCESS, FAILURE"

(I don't know which server these came from though, it's possible it's not O365, but still, in the current implementation of Aiosmtpd, the email would have been refused).

As for O365, we did had a few issues in the past too. If you rewrite the email using the Python library, some might break, they are strict on the headers, even the "Received" one. But even though they are difficult to work with, I kind of like it : they enforce a proper structure to respect and force developers to adhere to the standard more. Truth is, if your email works at Outlook, it should work everywhere else too.

**Comment by leandroltavares** [comment]
Hey guys, any updates on this? I wonder if the addition of the `hook` is planned to be added and released?

**Comment by nim-odoo** [comment]
If I understand correctly https://www.rfc-editor.org/rfc/rfc4954#section-5

> For this reason, servers that advertise support for this
> extension MUST support the AUTH parameter to the MAIL FROM
> command even when the client has not authenticated itself to the
> server.

Since `aiosmtpd` advertises `AUTH LOGIN PLAIN` by default, it should not reject mails using `AUTH` in their `MAIL FROM`.

**Comment by nim-odoo** [comment]
Still early, but it seems the following fixes the issue related to `AUTH=<>`:

```py
    async def handle_EHLO(self, server, session, envelope, hostname, response):
        session.host_name = hostname
        return [r for r in response if not r.startswith("250-AUTH")]
```

But of course it only works if clients properly implement the RFC.

**Comment by adbenitez** [comment]
I think I am having this problem while forwarding incoming email from postfix to a script that uses `aiosmtpd`, these are the logs:

![Image](https://github.com/user-attachments/assets/5053fc49-2cbf-4817-8ef3-5037dc9f98ef)

**Comment by cnicodeme** [comment]
I ended up extending the `SMTP` class by implementing my own custom `mail_from` function that was a basic copy of the original, but ignores the other given parameters (like `auth`)

**Comment by adbenitez** [comment]
@cnicodeme could you provide the class? I would also like to use it

**Comment by nim-odoo** [comment]
> [@cnicodeme](https://github.com/cnicodeme) could you provide the class? I would also like to use it

We have been using this for 2 years without issue:
```py
class MySmtp:
...
    async def handle_EHLO(self, server, session, envelope, hostname, response):
        session.host_name = hostname
        return [r for r in response if not r.startswith("250-AUTH")]
...
```

**Comment by adbenitez** [comment]
@nim-odoo  will that help in my situation tho? (see my screenshot) I think it is not related to AUTH, also not related to O365

**Comment by cnicodeme** [comment]
We simply created an extended class (`class CustomSMTP(SMTP)`), copied the function from `SMTP` and commented out the part posing issue on the SMTP Class:

```python
class CustomSMTP(SMTP):
    @syntax('MAIL FROM: <address>', extended=' [SP <mail-parameters>]')
    async def smtp_MAIL(self, arg: Optional[str]) -> None:
        if await self.check_helo_needed():
            return
        if await self.check_auth_needed("MAIL"):
            return
        syntaxerr = '501 Syntax: MAIL FROM: <address>'
        assert self.session is not None
        if self.session.extended_smtp:
            syntaxerr += ' [SP <mail-parameters>]'
        if arg is None:
            await self.push(syntaxerr)
            return
        arg = self._strip_command_keyword('FROM:', arg)
        if arg is None:
            await self.push(syntaxerr)
            return
        address, addrparams = self._getaddr(arg)
        if address is None:
            await self.push("553 5.1.3 Error: malformed address")
            return
        if not address:
            await self.push(syntaxerr)
            return
        if not self.session.extended_smtp and addrparams:
            await self.push(syntaxerr)
            return
        assert self.envelope is not None
        if self.envelope.mail_from:
            await self.push('503 Error: nested MAIL command')
            return
        assert addrparams is not None
        mail_options = addrparams.upper().split()
        params = self._getparams(mail_options)
        if params is None:
            await self.push(syntaxerr)
            return
        if not self._decode_data:
            body = params.pop('BODY', '7BIT')
            if body not in ['7BIT', '8BITMIME']:
                await self.push(
                    '501 Error: BODY can only be one of 7BIT, 8BITMIME')
                return
        smtputf8 = params.pop('SMTPUTF8', False)
        if not isinstance(smtputf8, bool):
            await self.push('501 Error: SMTPUTF8 takes no arguments')
            return
        if smtputf8 and not self.enable_SMTPUTF8:
            await self.push('501 Error: SMTPUTF8 disabled')
            return
        self.envelope.smtp_utf8 = smtputf8
        size = params.pop('SIZE', None)
        if size:
            if isinstance(size, bool) or not size.isdigit():
                await self.push(syntaxerr)
                return
            elif self.data_size_limit and int(size) > self.data_size_limit:
                await self.push(
                    '552 Error: message size exceeds fixed maximum message '
                    'size')
                return
        """
        if len(params) > 0:
            await self.push(
                '555 MAIL FROM parameters not recognized or not implemented')
            return
        """
        status = await self._call_handler_hook('MAIL', address, mail_options)
        if status is MISSING:
            self.envelope.mail_from = address
            self.envelope.mail_options.extend(mail_options)
            status = '250 OK'
        log.info('%r sender: %s', self.session.peer, address)
        await self.push(status)
```

That's all ;)

**Comment by adbenitez** [comment]
@cnicodeme thanks a lot!

**Comment by cnicodeme** [comment]
Happy to help 😊

**Comment by adbenitez** [comment]
for future readers: in my case the unsupported parameter was coming via RCPT TO, so I also needed to do similar patch overriding `smtp_RCPT`  commenting the 555 response there as well
